File: server/app/services/leave_service.py
from datetime import datetime
from typing import List, Dict, Any
from fastapi import HTTPException
from sqlalchemy.orm import Session
from app.models.user import DoctorProfile
from app.models.appointment import Appointment
from app.models.leave import DoctorLeave
from app.services.email_service import email_service
from app.services.calendar_service import calendar_service
from app.utils.logger import logger

class LeaveService:
    def preview_conflicts(self, db: Session, doctor_id: str, start_date: str, end_date: str) -> Dict[str, Any]:
        conflicts = db.query(Appointment).filter(
            Appointment.doctor_id == doctor_id,
            Appointment.date >= start_date,
            Appointment.date <= end_date,
            Appointment.status == "CONFIRMED"
        ).all()

        return {
            "conflictCount": len(conflicts),
            "conflicts": [
                {
                    "id": a.id,
                    "patientName": a.patient.name,
                    "date": a.date,
                    "startTime": a.start_time,
                    "symptoms": a.symptoms
                }
                for a in conflicts
            ]
        }

    def apply_leave(self, db: Session, doctor_id: str, start_date: str, end_date: str, reason: str) -> DoctorLeave:
        doctor = db.query(DoctorProfile).filter(DoctorProfile.id == doctor_id).first()
        if not doctor:
            raise HTTPException(status_code=404, detail="Doctor profile not found")

        # Query conflicting appointments
        conflicting_appts = db.query(Appointment).filter(
            Appointment.doctor_id == doctor_id,
            Appointment.date >= start_date,
            Appointment.date <= end_date,
            Appointment.status == "CONFIRMED"
        ).all()

        # Update appointment statuses and notify patients
        for appt in conflicting_appts:
            appt.status = "CANCELLED_DOCTOR_LEAVE"
            appt.cancellation_reason = f"Physician scheduled leave: {reason}"

            if appt.google_calendar_event_id:
                calendar_service.delete_event(appt.google_calendar_event_id)

            email_service.send_leave_cancellation_notice(
                db,
                patient_email=appt.patient.email,
                patient_name=appt.patient.name,
                doctor_name=doctor.user.name,
                date=appt.date,
                start_time=appt.start_time,
                reason=reason
            )

        leave = DoctorLeave(
            doctor_id=doctor_id,
            start_date=start_date,
            end_date=end_date,
            reason=reason,
            status="APPROVED",
            affected_appointments_count=len(conflicting_appts)
        )
        db.add(leave)
        db.commit()
        db.refresh(leave)

        logger.info(f"🏖️ Doctor leave applied for {doctor.user.name} ({start_date} to {end_date}). Affected appointments: {len(conflicting_appts)}")
        return leave

leave_service = LeaveService()

File: server/src/services/leave.service.ts
import { prisma } from '../prisma.js';
import { emailService } from './email.service.js';
import { calendarService } from './calendar.service.js';
import { logger } from '../utils/logger.js';

export class LeaveService {
  /**
   * Applies leave for a doctor and automatically resolves conflicting appointments
   */
  public async applyDoctorLeave(params: {
    doctorId: string;
    startDate: string; // YYYY-MM-DD
    endDate: string; // YYYY-MM-DD
    reason: string;
  }) {
    // 1. Find all active confirmed appointments in this date range
    const conflictingAppointments = await prisma.appointment.findMany({
      where: {
        doctorId: params.doctorId,
        date: {
          gte: params.startDate,
          lte: params.endDate,
        },
        status: 'CONFIRMED',
      },
      include: {
        patient: true,
        doctor: { include: { user: true } },
      },
    });

    const affectedCount = conflictingAppointments.length;

    // 2. Create the DoctorLeave record
    const leave = await prisma.doctorLeave.create({
      data: {
        doctorId: params.doctorId,
        startDate: params.startDate,
        endDate: params.endDate,
        reason: params.reason,
        status: 'APPROVED',
        affectedAppointmentsCount: affectedCount,
      },
      include: {
        doctor: { include: { user: true } },
      },
    });

    // 3. Atomically update conflicting appointments and notify patients
    if (affectedCount > 0) {
      logger.warn(`Doctor leave triggered conflict resolution for ${affectedCount} appointments (Dr. ${leave.doctor.user.name})`);

      for (const appt of conflictingAppointments) {
        // Update appointment status
        await prisma.appointment.update({
          where: { id: appt.id },
          data: {
            status: 'CANCELLED_DOCTOR_LEAVE',
            cancellationReason: `Doctor on approved leave: ${params.reason}`,
          },
        });

        // Release/Delete Google Calendar event
        calendarService.deleteEvent({
          userId: appt.patientId,
          eventId: appt.googleCalendarEventId,
        }).catch(() => {});

        // Dispatch High-Priority Email to Patient
        emailService.sendDoctorLeaveCancellation({
          patientEmail: appt.patient.email,
          patientName: appt.patient.name,
          doctorName: appt.doctor.user.name,
          date: appt.date,
          startTime: appt.startTime,
          reason: params.reason,
        }).catch((err) => logger.error(`Failed to send leave cancellation email to ${appt.patient.email}:`, err));
      }
    }

    // 4. Also release any active slot holds in that range
    await prisma.slotHold.updateMany({
      where: {
        doctorId: params.doctorId,
        date: {
          gte: params.startDate,
          lte: params.endDate,
        },
        status: 'ACTIVE',
      },
      data: { status: 'RELEASED' },
    });

    return {
      leave,
      affectedAppointmentsCount: affectedCount,
      affectedAppointments: conflictingAppointments.map((a) => ({
        id: a.id,
        patientName: a.patient.name,
        patientEmail: a.patient.email,
        date: a.date,
        startTime: a.startTime,
      })),
    };
  }

  /**
   * Previews conflicts for a proposed leave before submission
   */
  public async previewLeaveConflicts(params: {
    doctorId: string;
    startDate: string;
    endDate: string;
  }) {
    const conflictingAppointments = await prisma.appointment.findMany({
      where: {
        doctorId: params.doctorId,
        date: {
          gte: params.startDate,
          lte: params.endDate,
        },
        status: 'CONFIRMED',
      },
      include: {
        patient: { select: { id: true, name: true, email: true } },
      },
      orderBy: [{ date: 'asc' }, { startTime: 'asc' }],
    });

    return {
      conflictCount: conflictingAppointments.length,
      conflicts: conflictingAppointments,
    };
  }

  /**
   * Retrieves leave history for a doctor
   */
  public async getDoctorLeaves(doctorId: string) {
    return prisma.doctorLeave.findMany({
      where: { doctorId },
      orderBy: { startDate: 'desc' },
    });
  }

  /**
   * Cancels/deletes a leave
   */
  public async deleteLeave(leaveId: string, doctorId?: string) {
    const whereClause: any = { id: leaveId };
    if (doctorId) whereClause.doctorId = doctorId;

    return prisma.doctorLeave.delete({
      where: whereClause,
    });
  }
}

export const leaveService = new LeaveService();
